### PR TITLE
Add `SubdiagnosticMessageOp` as a trait alias.

### PR DESCRIPTION
**Pull request by nnethercote**
It avoids a lot of repetition.

r? matthewjasper

### CODE REVIEW

File: compiler/rustc_errors/src/diagnostic.rs [hunk 1]
```diff
@@ -77,11 +77,12 @@ where
 
     /// Add a subdiagnostic to an existing diagnostic where `f` is invoked on every message used
     /// (to optionally perform eager translation).
-    fn add_to_diagnostic_with<F>(self, diag: &mut Diagnostic, f: F)
-    where
-        F: Fn(&mut Diagnostic, SubdiagnosticMessage) -> SubdiagnosticMessage;
+    fn add_to_diagnostic_with<F: SubdiagnosticMessageOp>(self, diag: &mut Diagnostic, f: F);
 }
 
+pub trait SubdiagnosticMessageOp =
```
Comment (compiler-errors) [inline compiler/rustc_errors/src/diagnostic.rs:83]: Note: This could also just be expressed as:

```rust
trait SubdiagnosticMessageOp: Fn(&mut Diagnostic, SubdiagnosticMessage) -> SubdiagnosticMessage {}
impl<F> SubdiagnosticMessageOp for F where F: Fn(&mut Diagnostic, SubdiagnosticMessage) -> SubdiagnosticMessage {}
```

SO I don't believe this commits us to any particularities of the `trait_alias` feature yet.

File: compiler/rustc_resolve/src/late/diagnostics.rs [hunk 2]
```diff
@@ -1151,7 +1151,7 @@ impl<'a: 'ast, 'ast, 'tcx> LateResolutionVisitor<'a, '_, 'ast, 'tcx> {
 
     fn suggest_changing_type_to_const_param(
         &mut self,
-        err: &mut Diagnostic,
+        err: &mut DiagnosticBuilder<'_>,
```
Comment (compiler-errors) [inline compiler/rustc_resolve/src/late/diagnostics.rs:1154]: what does this have to do with anything?

**Review by compiler-errors — APPROVED** [review]
Please either explain or revert that stray change, r=me after

File: compiler/rustc_errors/src/diagnostic.rs (hunk 1 above)
Comment (nnethercote) [reply compiler/rustc_errors/src/diagnostic.rs:83]: There's one other use of `trait_alias` in the compiler so I figured it was fair game. I think this comment is expressing the opinion that using `trait alias` is ok?

File: compiler/rustc_resolve/src/late/diagnostics.rs (hunk 2 above)
Comment (nnethercote) [reply compiler/rustc_resolve/src/late/diagnostics.rs:1154]: Yeah, accidental inclusion from some follow-up changes, I will revert it.

File: compiler/rustc_errors/src/diagnostic.rs (hunk 1 above)
Comment (compiler-errors) [reply compiler/rustc_errors/src/diagnostic.rs:83]: Yes, it's ok

**Comment by nnethercote** [comment]
I removed the unintended change.

@bors r=compiler-errors
